### PR TITLE
Comms expansion

### LINKED ISSUE #15: Future functionalities for communication handler
# Features
These are the basic features to implement in the future.
## Desired Behavior / Functionality
* Modelling different communication technologies (optical, radio)
 * Modelling multiple communication bands 
 * Modelling finite transmission and receiving radiation diagrams (TBD) 
 * Checking if the time is the same for all the actors 
 * Handling scenarios where multiple actors A,B wants to communicate with the same actor C
 * Modelling communications disconnections and loss of packets
 
# Advanced Features
* Modelling different modulations (this affects the number of bits that you averagely lose and the effective bandwidth that you can use)
* Modelling Adaptive Modulations

### PR DESCRIPTION
**Pull request by timsmitdelft**
# Description

Summary of changes

* Added link budget modelling for communication, optical and radio. 
  * Added models for transmitters, receivers and links
  * Integrated transmitters and receivers into existing SpacecraftActors and GroundstationActors
  * Integrated links into existing PASEOS instances.
* Added Python notebooks to test/show satellite to ground (radio) and intersatellite (optical) 

## Resolved Issues

- [ ] fixes #15 

## How Has This Been Tested?
Calculations have been tested by comparing with papers and SMAD.
Software models have been tested by creating unit tests.

### CODE REVIEW

**Comment by GabrieleMeoni** [comment]
@gomezzz.  I cannot ask for your review, but it would be good if you could also take a look at this PR.

File: paseos/communication/radio_receiver_model.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+        self.polarization_loss = polarization_loss  # in dB
+        self.noise_temperature = 10 * math.log10(noise_temperature)  # in dBK
+
+    def set_gain(self, wavelength: float = 0) -> None:
+        """Sets gain for a receiver, based on the given gain, or antenna diameter and wavelength.
+
+        Args:
+            wavelength (float): The wavelength of the link, in meters
+        """
+        if self.antenna_gain == 0:
```
Comment (gomezzz) [inline paseos/communication/radio_receiver_model.py:47]: is None instead of == 0 default value check

File: paseos/communication/radio_transmitter_model.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+            point_losses,
+            antenna_gain,
+            antenna_diameter,
+        )
+        assert (
+                antenna_gain > 0 or antenna_diameter > 0
+        ), "Antenna gain or antenna diameter needs to be higher than 0."
+        assert not (antenna_diameter != 0 and antenna_gain != 0), (
+            "Only set one of antenna gain and " "antenna diameter, not both."
+        )
```
Comment (gomezzz) [inline paseos/communication/radio_transmitter_model.py:51]: as above

File: paseos/communication/radio_transmitter_model.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+        logger.debug("Initializing radio transmitter model.")
+
+    def set_gain(self, wavelength: float = 0) -> None:
+        """Sets gain for a transmitter, based on the given gain, or antenna diameter and wavelength.
+
+        Args:
+            wavelength (float): The wavelength of the link, in meters
+        """
+        if self.antenna_gain == 0:
```
Comment (gomezzz) [inline paseos/communication/radio_transmitter_model.py:65]: as above

File: paseos/communication/receiver_model.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+            "Only set one of antenna gain and " "antenna diameter, not both."
+        )
+        logger.debug("Initializing receiver model.")
+
+        self.line_losses = line_losses  # in dB
+        self.antenna_diameter = antenna_diameter  # in m
+        self.antenna_gain = antenna_gain  # in dB
+
+    def set_gain(self, wavelength: float = 0) -> None:
+        pass
```
Comment (gomezzz) [inline paseos/communication/receiver_model.py:34]: should this ever be called? if not, then I would suggest raising a NotImplementedError.

File: paseos/communication/transmitter_model.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+            antenna_diameter (float): The diameter of the antenna, in m. Either this or the gain
+            needs to be given.
+        """
+        assert input_power > 0, "Input power needs to be higher than 0."
+        assert 0 < power_efficiency <= 1, "Power efficiency should be between 0 and 1."
+        assert (
+                0 < antenna_efficiency <= 1
+        ), "Antenna efficiency should be between 0 and 1."
+        assert line_losses >= 0, "Line losses needs to be 0 or higher."
+        assert point_losses >= 0, "Pointing losses needs to be 0 or higher."
```
Comment (gomezzz) [inline paseos/communication/transmitter_model.py:38]: as a general comment, I would suggest to move most input value checks into the actor_builder. this is how it is done for the rest of the code and it can save you a lot of code duplication in your constructors

File: paseos/communication/transmitter_model.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    def set_EIRP(self) -> None:
+        """Sets the Effective Isotropic Radiated Power (EIRP) for a transmitter."""
+        self.EIRP = (
+                self.output_power - self.line_losses - self.point_losses + self.antenna_gain
+        )
+
+    def set_gain(self) -> None:
+        pass
+
+    def set_active(self, state: bool) -> None:
```
Comment (gomezzz) [inline paseos/communication/transmitter_model.py:58]: would either call the variable `is_active` and/or make the functions 
```py
def set_active():
    self.is_active = True
def set_inactive():
    self.is_active = False

```

`set_active(state=False)` is confusing

File: paseos/paseos.py [hunk 7]
```diff
@@ -27,6 +27,9 @@ class PASEOS:
     # The actor of the device this is running on
     _local_actor = None
 
+    # List of communication links
+    _communication_links = None
```
Comment (gomezzz) [inline paseos/paseos.py:31]: as mentioned, I would suggest an actor is responsible to maintain its own comm links. At the moment, if you want to run actors on different devices (not sharing memory , using e.g. MPI) you can simply exchange actors between devices. Having this here is very prone to error, since you could forget to manually also share the comm links at it is the only thing not contained within the actor

File: paseos/tests/communication_test.py [hunk 8]
```diff
@@ -0,0 +1,611 @@
...
+
+    ActorBuilder.add_comm_device(
+        actor=maspalomas_groundstation,
+        device_name=receiver_name,
+        noise_temperature=135,
+        line_losses=1,
+        polarization_losses=3,
+        antenna_gain=62.6,
+        device_type=DeviceType.RADIO_RECEIVER,
+    )
```
Comment (gomezzz) [inline paseos/tests/communication_test.py:35]: if you have some documentation for these values / how they are chosen would be good to put it here

File: paseos/tests/communication_test.py [hunk 9]
```diff
@@ -0,0 +1,611 @@
...
+from paseos.communication.optical_receiver_model import OpticalReceiverModel
+from paseos.communication.optical_transmitter_model import OpticalTransmitterModel
+from paseos.communication.radio_link_model import RadioLinkModel
+from paseos.communication.radio_receiver_model import RadioReceiverModel
+from paseos.communication.radio_transmitter_model import RadioTransmitterModel
+from paseos.communication.receiver_model import ReceiverModel
+from paseos.communication.transmitter_model import TransmitterModel
+
+
+def test_link_creation():
```
Comment (gomezzz) [inline paseos/tests/communication_test.py:20]: please add a short description what this test does

File: paseos/tests/communication_test.py [hunk 10]
```diff
@@ -0,0 +1,611 @@
...
+    )
+    assert isinstance(optical_link, OpticalLinkModel)
+
+    radio_link = RadioLinkModel(
+        sat_actor, radio_name, maspalomas_groundstation, receiver_name, 8675e6
+    )
+    assert isinstance(radio_link, RadioLinkModel)
+
+
+def test_receiver_creation():
```
Comment (gomezzz) [inline paseos/tests/communication_test.py:124]: pls provide a short description what the test does

File: paseos/communication/radio_transmitter_model.py [hunk 11]
```diff
@@ -0,0 +1,70 @@
+import math
+
+from loguru import logger
+
+from .transmitter_model import TransmitterModel
+from ..utils.gain_calc import calc_radio_gain_from_wavelength_diameter
+
+
+class RadioTransmitterModel(TransmitterModel):
```
Comment (gomezzz) [inline paseos/communication/radio_transmitter_model.py:9]: As a general (and maybe educational aspect), in terms of software design, I thin,k you are overusing classes and inheritance. The majority of the code in each subclass is a duplication of code in other files, the functionality of the different classes is quite similar and the created classes are not really necessary since you are not actually introducing any state (meaning the classes don't change over the course of the simulation / behave differently).

You could vastly simplify the code by storing the (constant) parameters of the comm link / transmitter / receiver in a dictionary inside the actor, then just pass the actor to the function that computes bitrate?

File: paseos/tests/communication_test.py [hunk 12]
```diff
@@ -0,0 +1,611 @@
...
+        device_name=optical_receiver_name,
+        line_losses=4.1,
+        antenna_gain=114.2,
+        device_type=DeviceType.OPTICAL_RECEIVER,
+    )
+
+    with pytest.raises(
+            AssertionError,
+            match="An optical transmitter is required for this optical link.",
+    ):
```
Comment (gomezzz) [inline paseos/tests/communication_test.py:82]: Why are these raised everywhere? This seems to indicate a problem in the structure to me if by default your tests raise exception that you have to capture

File: paseos/tests/link_budget_test.py [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+            AssertionError, match="Antenna efficiency should be between 0 and 1."
+    ):
+        calc_radio_gain_from_wavelength_diameter(10, 2, -0.5)
+    with pytest.raises(
+            AssertionError, match="Antenna efficiency should be between 0 and 1."
+    ):
+        calc_radio_gain_from_wavelength_diameter(10, 2, 1.5)
+
+    gain_radio = calc_radio_gain_from_wavelength_diameter(0.035, 15, 0.5)
+    assert np.isclose(gain_radio, 59.6, rtol=0.01, atol=1.0)
```
Comment (gomezzz) [inline paseos/tests/link_budget_test.py:29]: did you compute these by hand are checking against them or how where they determined?

File: paseos/utils/gain_calc.py
Comment (gomezzz) [inline paseos/utils/gain_calc.py:1]: Would move into the communication folder since they are only relevant for comms

File: paseos/utils/link_budget_calc.py [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+
+from .sky_field_sky_coordinate import SkyfieldSkyCoordinate
+from ..actors.ground_station_actor import GroundstationActor
+from ..actors.spacecraft_actor import SpacecraftActor
+
+_SKYFIELD_EARTH_PATH = os.path.join(
+    os.path.dirname(__file__) + "/../resources/", "de421.bsp"
+)
+# Skyfield Earth, in the future we may not always want to load this.
+_SKYFIELD_EARTH = load(_SKYFIELD_EARTH_PATH)["earth"]
```
Comment (gomezzz) [inline paseos/utils/link_budget_calc.py:15]: I believe this is already loaded in another file with the same name so this may create some weird interactions? (or at least load it twice). We might want to move it to the constants.py or similar?

File: paseos/utils/link_budget_calc.py [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+
+
+def calc_dist_and_alt_angle_spacecraft_ground(
```
Comment (gomezzz) [inline paseos/utils/link_budget_calc.py:18]: are these functions ever tested? Would at tests for them

**Review by gomezzz — CHANGES_REQUESTED** [review]
A few comments :) Overall quite nice! 

Two main concerns for me atm:

1. Please add a minimal example to the README.MD on how to use this
2. See the longer comment in one of the models, you are using a lot of different classes that are very similar. I would suggest to simplify that

File: examples/Communication_example/communication_example_utils.py [hunk 16]
```diff
@@ -0,0 +1,82 @@
+import numpy as np
+import pandas as pd
+
+
+def get_known_actor_comms_status(values):
+    """Helper function to track comms status"""
```
Comment (GabrieleMeoni) [inline examples/Communication_example/communication_example_utils.py:6]: Inputs shall be described in the docstring. You can check other files for examples.
(e.g., values (float): values are....
Returns: 
      return_name : what it is.

Please do the same for the other functions.

File: examples/Communication_example/communication_example_utils.py [hunk 17]
```diff
@@ -0,0 +1,82 @@
...
+    conv_values = []
+    for val in values:
+        status = ["No signal", "Ground only", "CommSat only", "Ground + Sat"]
```
Comment (GabrieleMeoni) [inline examples/Communication_example/communication_example_utils.py:9]: status shall be implemented as enum, not as string.

File: examples/Communication_example/communication_example_utils.py [hunk 18]
```diff
@@ -0,0 +1,82 @@
...
+    """Helper function to distance status history"""
+    return link.distance_history
+
+
+def get_elevation_angle_status(link):
+    """Helper function to get elevation angle status history"""
+    return link.elevation_angle_history
+
+
+def get_closest_entry(df, t, id):
```
Comment (GabrieleMeoni) [inline examples/Communication_example/communication_example_utils.py:35]: Docstring missing.

File: examples/Communication_example/communication_example_utils.py [hunk 19]
```diff
@@ -0,0 +1,82 @@
...
+    df_id = df[df.ID == id]
+    return df_id.iloc[(df_id["Time"] - t).abs().argsort().iloc[:1]]
+
+
+def get_analysis_df(df, timestep=60, orbital_period=1):
```
Comment (GabrieleMeoni) [inline examples/Communication_example/communication_example_utils.py:40]: Docstring missing.

File: paseos/communication/link_model.py [hunk 20]
```diff
@@ -0,0 +1,128 @@
...
+
+from .receiver_model import ReceiverModel
+from .transmitter_model import TransmitterModel
+from ..actors.base_actor import BaseActor
+
+
+class LinkModel:
+    """This class defines a link model, containing one transmitter and one receiver."""
+
+    _bitrate_history = []
```
Comment (GabrieleMeoni) [inline paseos/communication/link_model.py:13]: Please comment each internal variable.

File: paseos/communication/link_model.py [hunk 21]
```diff
@@ -0,0 +1,128 @@
...
+        self._bitrate_history = []
+        self._line_of_sight_history = []
+        self._distance_history = []
+        self._elevation_angle_history = []
+
+    def get_path_loss(self, slant_range: float) -> float:
+        """Gets the path loss (free space loss) for a link.
+
+        Args:
+            slant_range (float): The slant range of the link, in meters
```
Comment (GabrieleMeoni) [inline paseos/communication/link_model.py:61]:
```suggestion
            slant_range (float): The slant range of the link, in meters
```
```suggestion
            slant_range (float): The slant range of the link, in meters.
```

File: paseos/communication/link_model.py [hunk 22]
```diff
@@ -0,0 +1,128 @@
...
+
+        Returns:
+            The path loss (free space loss) in dB
```
Comment (GabrieleMeoni) [inline paseos/communication/link_model.py:64]:
```suggestion
            The path loss (free space loss) in dB.
```

File: paseos/communication/optical_link_model.py [hunk 23]
```diff
@@ -0,0 +1,107 @@
...
+
+        assert isinstance(transmitter, OpticalTransmitterModel), (
+            "An optical transmitter is required " "for this optical link."
+        )
+        assert isinstance(
+            receiver, OpticalReceiverModel
+        ), "An optical receiver is required for this optical link."
+
+        logger.debug("Initializing optical link model.")
+        self.required_BER = 10e-3
```
Comment (GabrieleMeoni) [inline paseos/communication/optical_link_model.py:51]: Why 10-3? I would have this value as user-defined value. In addition, this is typically linked with the maximum data-rate that you can have.

File: paseos/communication/optical_link_model.py [hunk 24]
```diff
@@ -0,0 +1,107 @@
...
+
+        self.modulation_scheme = "OOK"
+        self.required_s_n_margin = 3  # in dB
```
[the remark this answers is not in the document]
Comment (GabrieleMeoni) [reply paseos/communication/optical_link_model.py:54]: Is S_N_margin linked to the desired BER? If yes, we should avoid the definition of two related constants.

File: paseos/communication/optical_link_model.py [hunk 25]
```diff
@@ -0,0 +1,107 @@
...
+
+        Returns:
+            The bitrate in bps
+        """
+        assert slant_range > 0, "Slant range needs to be higher than 0 meters"
+        assert min_elevation_angle >= 0, "Slant range needs to be higher than 0 meters"
+
+        self.total_channel_loss = self.get_path_loss(slant_range)
+
+        self.signal_at_receiver = self.transmitter.EIRP - self.total_channel_loss
```
[the remark this answers is not in the document]
Comment (GabrieleMeoni) [reply paseos/communication/optical_link_model.py:87]: Are those values on a logarithmic scale? If yes, please, specify it in the names (e.g., EIRP_dB or something like that).

[the remark this answers is not in the document]
Comment (GabrieleMeoni) [reply paseos/communication/optical_link_model.py:87]: If these are in logarithmic scales, that means the loss of your signal will reduce your power to 0 dB. The effect of that depends on the noise power. If the noise power is higher than 0 dB, then you will have a terrible Bit Error Rate.

File: paseos/communication/radio_link_model.py [hunk 26]
```diff
@@ -0,0 +1,127 @@
...
+
+        logger.debug("Initializing radio link model.")
+        self.transmitter = transmitter
+        self.receiver = receiver
+        self.required_BER = 10e-5
+        self.frequency = frequency  # in Hz
+        self.modulation_scheme = "BPSK"
+        self.zenith_atmospheric_attenuation = 0.5  # in dB
+        self.required_s_n_ratio = 9.6  # in dB
+        self.required_s_n_margin = 3  # in dB
```
[the remark this answers is not in the document]
Comment (GabrieleMeoni) [reply paseos/communication/radio_link_model.py:56]: Again, I  would let the user specify margins and the other parameters.

File: paseos/tests/link_budget_test.py (hunk 13 above)
Comment (timsmitdelft) [reply paseos/tests/link_budget_test.py]: Yes, are computed by hand. Also added a comment to clarify that.

File: paseos/tests/communication_test.py (hunk 12 above)
Comment (timsmitdelft) [reply paseos/tests/communication_test.py]: A number of these specic assertions are now removed because of imrpoved code structure. The rest are testing the asserts in ActorBuilder. Checking that are errors are raised when parameters such as input power are set as negative.

File: paseos/communication/radio_transmitter_model.py (hunk 11 above)
Comment (timsmitdelft) [reply paseos/communication/radio_transmitter_model.py]: I agree that I made it slightly to complicated with the receiver, transmitter and link subclasses. Simplified that to simply ReceiverModel, TransmitterModel and LinkModel. I guess these could be a dictionary inside the actor, though the models do keep track of their active state and the bitrate histories. Additionally, I find this a lot cleaner than a large dictionary or DotMap.

File: paseos/utils/link_budget_calc.py (hunk 14 above)
Comment (timsmitdelft) [reply paseos/utils/link_budget_calc.py]: Moved it to constants.py

File: paseos/communication/optical_link_model.py (hunk 23 above)
Comment (timsmitdelft) [reply paseos/communication/optical_link_model.py]: For optical this the BER for which the equation to determine bitrate is valid. It was quite difficult to find an equation where the bitrate could be calculated from. The one I could find so far is only valid for BER 10E-3 and a wavelength of 1550 nm.

**Comment by timsmitdelft** [comment]
Opened issue #210 for some possible improvements in the future.

**Comment by gomezzz** [comment]
Closing for now since I don't think this is being actively worked on, please free to re-open ✌️
